ocfweb/docs/docs/staff/backend/kubernetes.md

```diff
@@ -0,0 +1,190 @@
+
+1. Leader elections in case of failure.
+2. Log replication across all masters.
+3. Ensuring log integrity across all masters.
+
+### Why the odd number of masters?
+
+Consider a cluster of *N* members. When masters form quorum to agree on cluster
+state, quorum must have _at least_ ⌊*N*/2⌋+1 members. Every new odd number in a
+cluster with *M* > 1 masters adds one more node of fault tolerance.  Therefore,
+adding an extra node to an odd numbered cluster gives us nothing. If interested
+read more [here][failure-tolerance].
+
+### Workers
+
+Workers are the brawn in the Kubernetes cluster. While master nodes are
+constantly sharing data, managing the control plane, and scheduling services,
+workers primarily run [pods][pod]. `kublet` is the service that executes pods
+as dictated by the control plane, performs health checks, and recovers from pod
+failures should they occur. Workers also run an instance of `kube-proxy`, which
+forwards control plane traffic to the correct `kublet`.
+
+### Pods
+
+In the Kubernetes world, pods are the smallest computing unit. A pod is made up
+of one or more containers. The difference between a pod and a standalone
+container is best illustrated by an example. Consider [ocfweb][ocfweb]; it is
+comprised of several containers—the web container, static container, and worker
+container.  In Kubernetes, together these three containers form one pod, and it
+is pods that can be scaled up or down. A failure in any of these containers
+indicates a failure in the entire pod. It should be noted that when writing
+Kubernetes services we don't actually deal in pods but one further abstraction,
+deployments, which create pods for us.
+
+## OCF Kubernetes Cluster Bootstrapping
+
+Since almost all OCF architecture is bootstapped using Puppet, it was necessary
+for us to do the same with Kubernetes. We rely on the
+[puppetlabs-kubernetes][kubernetes-module] module to handle initial
+bootstrapping and bolt OCF specific configurations on top of it.
+`puppetlabs-kubernetes` performs two crucial tasks:
+
+1. Installs `etcd`, `kubelet`, `kube-proxy`, and `kube-dns`, initializes the
+   cluster, and applies a networking backend.
+2. Generates the required [PKI for Kubernetes and etcd][kubernetes-pki].
+
+Do note that `puppetlabs-kubernetes` is still very much a work in progress. If
+you notice an issue in the module you are encouraged to write a patch and send
+it upstream.
+
+### Kubernetes PKI
+
+All the private keys and certs for the PKI are in the puppet private share, in
+`/opt/puppet/shares/private/kubernetes`. We won't go into detail of everything
+contained there, but much of Kubernetes and `etcd` communication is
+authenticated using client certificates. All the necessary items for workers
+are included in `os/Debian.yaml`, although adding a new master to the cluster
+requires re-running [kubetool][puppetlabs-kubetool] to generate new `etcd
+server` and `etcd peer` certs.
```

Comment:
I'd add that as far as we know, refreshing the etcd certs requires taking down the cluster.

An aside that we may or may not want to document-- I believe it should, in theory, be possible to do all of this live and without downtime, by generating and signing keys using the old CA. We'd have to see exactly what the kubetool script is doing. If we ever want to scale to 5 masters, this is something that could be worth looking into.